SAMZA-1126 - Semantics of processorId in Samza

navina opened the pull request:
Implementation based on [SEP-1](https://cwiki.apache.org/confluence/display/SAMZA/SEP-1%3A+Semantics+of+ProcessorId+in+Samza)

navina commented:
Fyi, primary implementation changes are in StreamProcessor.java, ProcessorIdGenerator.java, SamzaObjectMapper.java, JsonContainerModelMixin.java and JsonJobModelMixin.java. 
Everything else is a side-effect of changing processorId from int to String. 

Due to the large-scope of this change, I would prefer to get explicit approvals from folks owning certain components. I have listed them below. Please review the main classes listed above and the components listed under your name. Appreciate the extra pair of eyes to make sure I don't break existing behavior. 

@vjagadish1989 All classes related to YarnClusterResourceManager - AppMasterService, ApplicationMaster{Rest,Web}Servlet, ContainerAllocator, ContainerRunner etc. 

@jmakes All class related to TaskAssignmentManager, BalancingTaskNameGrouper implementation and tests.

@nickpan47 All classes :) 

Thanks a lot everyone!

nickpan47 commented on samza-core/src/test/scala/org/apache/samza/container/TestTaskInstance.scala:
```diff
@@ -48,7 +48,7 @@ import org.apache.samza.system.SystemAdmin
 
 class TestTaskInstance {
-  @Test
+ /* @Test
```
Same here. Please add comments and would prefer to move the changes to disable and fix the tests in a separate JIRA and PR.

navina replied:
Ah.. Yet another file I missed . Fixing it :)

navina commented on samza-api/src/main/java/org/apache/samza/runtime/ProcessorIdGenerator.java:
```diff
@@ -0,0 +1,36 @@
+@InterfaceStability.Evolving
+public interface ProcessorIdGenerator {
+  /**
+   * Generates a String representation to identify the processor instance
```
ProcessorId is a unique identifier for each StreamProcessor instance. not the jvm. 
May be I should rephrase document as : 
"Generates a String representation to identify an instance of {@link StreamProcessor}"

I can further add a note about the case where a single JVM has multiple StreamProcessor instance. 
Do you think that will clarify it better?

navina commented on samza-core/src/main/java/org/apache/samza/config/ApplicationConfig.java:
```diff
@@ -0,0 +1,58 @@
+ */
+public class ApplicationConfig extends MapConfig {
+  /**
+   * processor.id is equivalent to containerId in samza. It is a logical identifier used by Samza for a processor.
```
I think we are supporting it. 
>  I don't think the processor.id is equivalent to containerId.
I hope you are also referring to containerId as the logical containerId used by Samza, which is different from the physical containerId provided by Yarn. Why do you think they are not equivalent? If we modify our Yarn-based implementation today to bring up more than one instance of SamzaContainer within the same JVM, it would be equivalent to the use-case we are offering now. Are you saying they are not equivalent because we have a 1:1 mapping between JVM and SamzaContainer today?

navina replied:
> we should move other app scope config variables to this class as well, just as app.class, app.runner.class, etc. And if some of those variables are used in API classes (e.g. ApplicationRunner.fromConfig() or RuntimeEnvironment.fromConfig()), we may need to move this to samza-api package as well

I spoke with Xinyu about this. He said he would follow-up with moving configs to this class because I didn't have full context on the ApplicationRunner implementation.

nickpan47 replied:
@navina yes, what I meant is exactly the current 1:1 mapping between a JVM process and SamzaContainer instance. Right now, the SamzaContainer.main() is invoked when starting the JVM process, which mandates a 1:1 mapping here. Also, the integer container_id is set in the current container process's environment and YarnClusterResourceManager.java is referring to that as a single integer to identify the physical JVM process locally.

navina replied:
Yep. Got it. I will rephrase the documentation based on our discussion from morning. Thanks for clarifying! 👍

navina commented on samza-core/src/main/java/org/apache/samza/coordinator/JobCoordinator.java:
```diff
@@ -57,10 +57,11 @@
+   * @return String representing a unique logical processor ID
    */
-  int getProcessorId();
+  String getLocalProcessorId();
```
> JobCoordinator's APIs are only called within local StreamProcessor, and there is no use case to call JobCoordinator to get processIds other than the StreamProcessor wrapping the JobCoordinator

I actually never even thought about the method for getting all processorIds through the JobCoordinator interface. That will be internal to the JC implementation and will not be used by any other interacting component, such as the StreamProcessor. Hence, I didn't find the documentation be confusing. 
I will try to clarify in the documentation that this interface container methods that are required for interacting with StreamProcessor.

navina commented on samza-core/src/main/java/org/apache/samza/processor/SamzaContainerController.java:
```diff
@@ -60,16 +60,14 @@
       Map<String, MetricsReporter> metricsReporterMap) {
     this.executorService = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
-        .setNameFormat("p" + processorId + "-container-thread-%d").build());
+        .setNameFormat("container-thread-%d").build());
```
Mm.. Good point. I think I was simply re-factoring and didn't see the need for processorId except debug logging. But multiple processors within the same JVM will make it hard to differentiate since the threadpool is scoped within each StreamProcessor. 

Let me fix this. Thanks for pointing it out.

navina commented on samza-core/src/main/java/org/apache/samza/serializers/model/JsonContainerModelMixIn.java:
```diff
@@ -19,23 +19,28 @@
 public abstract class JsonContainerModelMixIn {
   @JsonCreator
-  public JsonContainerModelMixIn(@JsonProperty("container-id") int containerId, @JsonProperty("tasks") Map<TaskName, TaskModel> tasks) {
+  public JsonContainerModelMixIn(@JsonProperty("processor-id") String processorId, @JsonProperty("tasks") Map<TaskName, TaskModel> tasks) {
```
When we read the old JSON container model, we won't be using the mixin anymore. I changed it to use the de-serializer. If the containerModel contains processor-id, all is well. If it  doesn't contain processor-id, it will return null. We, then, check to see if container-id is defined. If it is, it is converted to a processor-id String. Otherwise, we throw an exception as we expect the Container Model to either contain container-id or processor-id. 
I was able to verify the deserialization with a unit test in TestSamzaObjectMapper. 
Let me know if you are talking about something else here.

nickpan47 replied:
👍 thanks for the explanation. Makes sense.

navina commented on samza-core/src/main/java/org/apache/samza/serializers/model/JsonJobModelMixIn.java:
```diff
@@ -30,12 +30,12 @@
 public abstract class JsonJobModelMixIn {
   @JsonCreator
-  public JsonJobModelMixIn(@JsonProperty("config") Config config, @JsonProperty("containers") Map<Integer, ContainerModel> containers) {
+  public JsonJobModelMixIn(@JsonProperty("config") Config config, @JsonProperty("containers") Map<String, ContainerModel> containers) {
```
I think jackson decoder handles integer/string conversions correctly. I remember trying out a simple test before making this change. 
Let me try to add a test so that it uses this Mixin. That way, we can all sleep peacefully :)

navina replied:
I verified the behavior of json parser in jackson library. Unless we configure json parser to "ALLOW_UNQUOTED_FIELD_NAMES", it will always serialize the field name as String and it will correctly deserialize it back to Integer/String as specified in the MixinAnnotation. 

For existing jobs, we can safely assume that containerId/processorId is always an integer value. So, it shouldn't be a problem to rollback or roll-forward.

nickpan47 replied:
👍 thanks for verifying! Super helpful.

navina commented on samza-core/src/main/java/org/apache/samza/serializers/model/SamzaObjectMapper.java:
```diff
@@ -89,10 +91,30 @@ public static ObjectMapper getObjectMapper() {
 
+    module.addDeserializer(ContainerModel.class, new JsonDeserializer<ContainerModel>() {
+      @Override
+      public ContainerModel deserialize(JsonParser jp, DeserializationContext ctxt) throws IOException, JsonProcessingException {
```
Rolling back shouldn't be a problem, right? Please note, that when we are serializing , we are writing both container-id and processor-id. Skipping one of them will make it hard to handle rollbacks. Does that address your concern?

nickpan47 replied:
Yep. That should be fine.

navina commented on samza-core/src/main/java/org/apache/samza/util/ClassLoaderHelper.java:
```diff
@@ -19,11 +19,30 @@
     return instance;
   }
+
+  public static <T> T fromClassName(String className, Class<T> classType) {
```
I thought this was the new util class for Class Loading. I am aware that ApplicationRunner also has similar usage. I wasn't sure if I should change it as a part of this patch. Let me know if you think otherwise.

nickpan47 replied:
I agree that we can put it in another PR, as we see more use cases like this.

navina commented on samza-core/src/main/java/org/apache/samza/zk/ZkCoordinationServiceFactory.java:
```diff
@@ -26,12 +26,10 @@
   synchronized public CoordinationUtils getCoordinationService(String groupId, String participantId, Config config) {
     ZkConfig zkConfig = new ZkConfig(config);
     ZkClient zkClient = new ZkClient(zkConfig.getZkConnect(), zkConfig.getZkSessionTimeoutMs(), zkConfig.getZkConnectionTimeoutMs());
-    ZkUtils zkUtils = new ZkUtils(participantId, new ZkKeyBuilder(groupId), zkClient, zkConfig.getZkConnectionTimeoutMs());
```
Good question. When I created ZkUtils, I wanted it to be a util class for interacting with Zookeeper. Hence, I wanted it to be stateless. 

It has somehow evolved into a stateful class (with processorId, which is what is being named as participantId). I have urged @sborya to put up a SEP for discussing the design of the CoordinationService because I don't fully agree with how it has been implemented. Due to lack of documentation, I am unable to answer all your questions with clarity.

So to answer your questions:

> What's this participantId?

Not sure why Boris called it that. But it seems to be the same as processorID.
> Are we changing the znode path after removing the participantId from the parameter? 
No. This "participantId" was only used in log statements.  It doesn't affect the coordination implementation in anyway.

All I wanted to make sure was that processorId is not being treated as a state in ZkUtils. Iirc, processorId was being used in the log statements only. Hence, I felt it is ok to remove it.

vjagadish1989 replied:
@nickpan47 @navina 
I remember I preferred (and still prefer) calling this `participantId` instead of `processorId` when I reviewed Boris's PR.  

It keeps the scope of this interface self-contained as a general purpose coordination API which is independent of `StreamProcessor` (or even `Samza` for that matter).

nickpan47 replied:
a) I would prefer to use processorId in the constructor parameter here, with more precise meaning of what's the participant in this particular context.
b) as long as the ZNode path is not affected by removing the input parameter here, it should be fine.

P.S. a side note to the ZkUtils class: usually, the list of entity names in the ZooKeeper methods are used to construct a znode path, which made me concerned about removing the participant parameter. Non-path related parameters usually occurs at the end.

navina replied:
> a) I would prefer to use processorId in the constructor parameter here, with more precise meaning of what's the participant in this particular context.

I would defer to @sborya  on whether to call it processorId or participantId. You can add this question here - [SAMZA-1173](https://issues.apache.org/jira/browse/SAMZA-1173)

> as long as the ZNode path is not affected by removing the input parameter here, it should be fine.

It shouldn't be affected.

> Non-path related parameters usually occurs at the end.
That's a good suggestion. Seems like a good pattern to follow :)

navina commented on samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java:
```diff
@@ -61,17 +65,24 @@
   private final CoordinationUtils coordinationUtils;
 
   private JobModel newJobModel;
-  private String newJobModelVersion;  // version published in ZK (by the leader)
```
I refactored it because it didn't need to be a member variable. However, since I don't have a good idea about this code, I think I will change this back. 

> Is it relevant to the processorId change?
No. Not related at all.

navina commented on samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java:
```diff
@@ -160,7 +158,7 @@ public void close() throws ZkInterruptedException {
     */
   public void subscribeToJobModelVersionChange(IZkDataListener dataListener) {
-    LOG.info("pid=" + processorId + " subscribing for jm version change at:" + keyBuilder.getJobModelVersionPath());
+    LOG.info(" subscribing for jm version change at:" + keyBuilder.getJobModelVersionPath());
```
getJobModelVersion Path returns the path to the znode where the version is stored. The paths are managed by ZkKeyBuilder. 

newJobModelVersion field was not used anywhere in the ZkJobCoordinator. As I explained above, I only refactored as it was not actually used by ZkJobCoordinator.

Let's sync-up in-person if you have further questions.

nickpan47 replied:
👍 to remove if newJobModelVersion is not used by ZkJobCoordinator. Wondering how it ends up in the code base if not used?

navina replied:
> Wondering how it ends up in the code base if not used?
It was probably missed in the review of the jobcoordinator PR?? Not sure Yi.

navina commented on samza-core/src/test/java/org/apache/samza/container/grouper/task/TestGroupByContainerCount.java:
```diff
@@ -163,31 +162,31 @@ public void testGroupManyTasks() {
    *  wasn't worth it. It certainly could be implemented in the future.
    */
-  @Test
+//  @Test - Not working - TODO: Fix it
```
Oh I fixed this test. Let me check and fix it. Thanks for pointing out.

navina commented on samza-core/src/test/java/org/apache/samza/container/grouper/task/TestGroupByContainerCount.java:
```diff
@@ -244,7 +243,7 @@ public void testBalancerAfterContainerIncrease() {
    *  because the scenario is infrequent, the benefits are not guaranteed, and the code complexity
    *  wasn't worth it. It certainly could be implemented in the future.
-   */
+   *//*
```
Ah.. these bunch of tests. I think I forgot to get to them after fixing the other million test failures. Let me work on this. Thanks for pointing out Yi! :)

navina commented on samza-core/src/test/java/org/apache/samza/container/grouper/task/TestGroupByContainerCount.java:
```diff
@@ -833,5 +832,5 @@ public void testBalancerResultImmutable() {
     Set<ContainerModel> containers = new GroupByContainerCount(2).balance(taskModels, localityManager);
     containers.remove(containers.iterator().next());
-  }
+  }*/
```
As stated above, this was not intentional. i will fix it.

navina commented on samza-core/src/test/java/org/apache/samza/task/TestAsyncRunLoop.java:
```diff
@@ -519,7 +519,7 @@ public void testEndOfStreamOffsetManagement() throws Exception {
   }
 
-  @Test
+//  @Test
```
This is part of the flaky test. This is being fixed by @santhoshvenkat1988  in [SAMZA-1183](issues.apache.org/jira/browse/SAMZA-1183)

navina commented on samza-core/src/test/java/org/apache/samza/zk/TestZkUtils.java:
```diff
@@ -161,7 +155,7 @@ public void handleDataDeleted(String dataPath)
   }
 
-  @Test
+/*  @Test
```
No. I just missed fixing this. Will take care of it now.

navina commented:
@nickpan47 I have addressed all your comments. Please let me know if anything pending.

@jmakes  @vjagadish1989 : Can you guys take a look at those classes I have assigned to you guys?

jmakes commented on samza-core/src/main/java/org/apache/samza/container/grouper/task/GroupByContainerCount.java:
```diff
@@ -74,7 +74,7 @@ public GroupByContainerCount(int containerCount) {
     Set<ContainerModel> containerModels = new HashSet<>();
     for (int i = 0; i < containerCount; i++) {
-      containerModels.add(new ContainerModel(i, taskGroups[i]));
+      containerModels.add(new ContainerModel(String.valueOf(i), taskGroups[i]));
```
This line, and I believe a few others, assume the container/processor ids are still a sequence of integers (even though the type is String). 

I see how this patch ensures that this code is only invoked in the legacy case, where that would be true. But this precondition/assumption is now subtle; it's not obvious by looking at the types. 

I suggest that we at least add a validation with a clear exception that ensures that the external strings (e.g. those that come from the TaskAssignmentManager) are all parseable integers.

jmakes replied:
A better long term approach is to completely overhaul these groupers to have no assumptions about the strings and treat them as arbitrary tokens.

navina replied:
> I suggest that we at least add a validation with a clear exception that ensures that the external strings (e.g. those that come from the TaskAssignmentManager) are all parseable integers.

What kind of validation would you suggest? In the Grouper itself? 

> A better long term approach is to completely overhaul these groupers to have no assumptions about the strings and treat them as arbitrary tokens

Yes. That was the intention with the standalone work. The task to replace with the new grouper has not yet been completed. I will add a TODO with a JIRA here so that we don't forget about it. For existing users running Yarn, this should continue to work as long as the AppMaster logic is not changed.

jmakes replied:
What kind of validation would you suggest? In the Grouper itself?
Yes, if the class expects all container/processor ids to be parsable as integers, it should have an explicit validation of all external processor ids with a message explaining that the grouper will not work with arbitrary strings like UUIDs. Then if someone mistakenly changes the code, they get a clear exception rather than strange grouping behavior.

navina replied:
@jmakes  This grouper essentially relies on job.container.count. It "generates" containerId according to the count and creates ContainerModels. There is no question of not being able to parse UUIDs. So, I am slightly lost of where the "external" processor ids are even passed in and where to add the validation. Let's discuss this in person and sort it out.

jmakes replied:
It "generates" some containerIds according to the count. It receives others from external sources like the task assignment manager. I was only suggesting that the latter should be validated s.t. we throw an exception rather than letting unexpected/undefined behavior occur.

vjagadish1989 commented on samza-core/src/main/java/org/apache/samza/coordinator/JobCoordinator.java:
```diff
@@ -55,12 +57,14 @@
-   * It is up to the user to ensure that different instances of StreamProcessor within a job have unique processor ID.
-   * @return integer representing the logical processor ID
+   * Returns the logical ID assigned to the processor that is local to the instance of StreamProcessor.
+   * It is the responsibility of the JobCoordinator to assign a unique identifier based on the underlying environment.
```
>> It is the responsibility of the JobCoordinator to assign a unique identifier based on the underlying environment

- Move it to the docs of the class instead of documenting it here? It we have already called out, this is probably redundant

- Will also be helpful to state the requirements of the ID here.

navina replied:
> Move it to the docs of the class instead of documenting it here? It we have already called out, this is probably redundant

Sure. Makes sense.

> Will also be helpful to state the requirements of the ID here.

Shouldn't requirements of the ID already stated in ProcessorIdGenerator?

vjagadish1989 commented on samza-core/src/main/java/org/apache/samza/job/model/ContainerModel.java:
```diff
@@ -37,32 +37,48 @@
+    this(containerId, Integer.parseInt(containerId), tasks);
+  }
+
+  public ContainerModel(String processorId, int containerId, Map<TaskName, TaskModel> tasks) {
```
Nit: Is it fair to say that this constructor will be deprecated?

navina replied:
Yeah.. I wasn't sure whether to annotate with deprecated because I introduced only now :) 
We should be deprecating this after 0.13, though.

vjagadish1989 commented on samza-core/src/main/java/org/apache/samza/job/model/ContainerModel.java:
```diff
@@ -87,6 +103,6 @@ public boolean equals(Object obj) {
 
   public int compareTo(ContainerModel other) {
-    return containerId - other.getContainerId();
+    return processorId.compareTo(other.getProcessorId());
```
Wondering if we should check for null processorId here and handle comparisons accordingly? That way, we don't crash with an NPE. The `hashCode` does seem to handle null values.

navina replied:
Hmm.. Yeah. Just read the compareTo documentation. We should be throwing NPE if either of them our null. Let me fix that and the documentation of the class, as recommended in Comparable interface

vjagadish1989 commented on samza-core/src/main/java/org/apache/samza/processor/SamzaContainerController.java:
```diff
@@ -44,7 +44,7 @@
 public class SamzaContainerController {
   private static final Logger log = LoggerFactory.getLogger(SamzaContainerController.class);
 
-  private final ExecutorService executorService;
```
Can a single `SamzaContainerController` manage lifecycle of multiple containers? Otherwise, we can take the processorId, and make the executorService immutable?

navina replied:
We may not manage multiple containers (at least, not as of today). However, based on the current ZkJobCoordinator implementation, we should be able to start and stop the container multiple times within the same lifecycle of the StreamProcessor. That said, this class can use some refactoring. I would like keep this outside the scope of this change. 
Ideally, I shouldn't have fixed this here either. So, I am going to revert back any unrelated changes in this class and fix them in the StreamProcessor/JobCoordinator refactoring.

vjagadish1989 commented on samza-core/src/main/java/org/apache/samza/processor/SamzaContainerController.java:
```diff
@@ -94,19 +90,21 @@ public SamzaContainerController(
   public void startContainer(ContainerModel containerModel, Config config, int maxChangelogStreamPartitions) {
```
Future clean-up: (pun intended ;) )

>> This method does not actually wait for the container to fully start-up
		
Would be nice to return a `Future` here. That way, the API looks more reasonable. Not sure what the scope of this will be. So, can defer it later.

navina replied:
Same answer as before. Let's focus on SEP-1 related changes alone here.

vjagadish1989 commented on samza-core/src/main/java/org/apache/samza/processor/SamzaContainerController.java:
```diff
@@ -60,16 +60,12 @@
-      String processorId,
       Map<String, MetricsReporter> metricsReporterMap) {
-    this.executorService = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
-        .setNameFormat("p" + processorId + "-container-thread-%d").build());
```
- What is the cardinality of the relationship between `SamzaContainerController` and `SamzaContainer`? 

Some potential opportunities for clean-up:
- It seems that invoking `startContainer()` multiple times can potentially clobber the saved state in this class. I suspect we can add a guard to prevent this.
- Very minor: Are n't we creating multiple instances of `executorService` but only stopping the last created instance in `shutdown()` if `startContainer` is invoked multiple times? 

These are probably not blockers, so feel free to defer these for a future clean-up.

navina replied:
As mentioned before, we have a separate task for refactoring.  Let's keep the clean-up in a separate patch. Thanks for the suggestions, though. Keep them coming :)

vjagadish1989 commented on samza-core/src/main/java/org/apache/samza/processor/SamzaContainerController.java:
```diff
@@ -44,7 +44,7 @@
 
-  private final ExecutorService executorService;
+  private ExecutorService executorService;
   private volatile SamzaContainer container;
```
Is this class thread-safe? If not, the `volatile` modifier is probably redundant here? If it is, there are other variables like `containerFuture` that are not thread-safe yet. Would be nice to fix one or both.

navina replied:
Good question. I would like to keep changes not associated with SEP-1 out of this PR.

vjagadish1989 commented on samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java:
```diff
@@ -83,51 +70,49 @@
-    updatedConfigMap.put(PROCESSOR_ID, String.valueOf(this.processorId));
-    Config updatedConfig = new MapConfig(updatedConfigMap);
+    // TODO: This check to be removed after 0.13+
+    ApplicationConfig applicationConfig = new ApplicationConfig(config);
```
Minor: Wondering if this check should be moved down to the `JobCoordinator` implementation instead of here?

navina replied:
Why do you think so? Since this check is only for existing use-cases of StreamProcessors, it seems acceptable to keep it here and removing it after 0.13+. If we move this to JobCoordinator abstraction layer, are you suggesting we move it to each and every implementation? 
If we decide to allow this check only for StandaloneJobCoordinator (because ZkJobCoordinator didn't exist before and hence, there is no question about `processor.id`), then the behavior is different across jobcoordinator implementation. We allow `processor.id` in Standalone versions in 0.13 and not in Zk versions in 0.13. That seems weird too. That's the reason why I have kept it at StreamProcessor level. 

This may be pushed to ApplicationRunner level though. However, ApplicationRunner is not yet fully fleshed out when I made these changes. We can move this check to other abstraction layers later on, if the need calls for it.

nickpan47 replied:
IMO, this would better be encapsulated in a loader method to get the ProcessorIdGenerator implementation class. No matter what implementation a JobCoordinator class takes, as long as it is loading ProcessorIdGenerator implementation, it should use the loader method and will perform this validation.

vjagadish1989 commented on samza-core/src/main/java/org/apache/samza/serializers/model/SamzaObjectMapper.java:
```diff
@@ -89,10 +91,30 @@ public static ObjectMapper getObjectMapper() {
     mapper.getSerializationConfig().addMixInAnnotations(JobModel.class, JsonJobModelMixIn.class);
     mapper.getDeserializationConfig().addMixInAnnotations(JobModel.class, JsonJobModelMixIn.class);
 
+    module.addDeserializer(ContainerModel.class, new JsonDeserializer<ContainerModel>() {
```
Minor: Slightly prefer being consistent with all other `DeSerializer`s. All others appear to define static inner classes as opposed to inlining them.

`  public static class ConfigDeserializer extends JsonDeserializer<Config> `

navina replied:
Sure

vjagadish1989 commented on samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java:
```diff
@@ -64,13 +64,11 @@
   private final int connectionTimeoutMs;
-  private final String processorId;
 
-  public ZkUtils(String processorId, ZkKeyBuilder zkKeyBuilder, ZkClient zkClient, int connectionTimeoutMs) {
```
Very minor: Slightly prefer `utils` classes as a collection of `static` methods instead of classes that contain state. Can rename this to `ZkClientWrapper` or something similar.

navina replied:
I prefer addressing non SEP-1 related suggestions in a separate PR. This is an awfully large change and should not be bloated with other important changes.

vjagadish1989 commented on samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java:
```diff
@@ -190,7 +188,7 @@ public void publishJobModel(String jobModelVersion, JobModel jobModel) {
    */
   public JobModel getJobModel(String jobModelVersion) {
-    LOG.info("pid=" + processorId + "read the model ver=" + jobModelVersion + " from " + keyBuilder.getJobModelPath(jobModelVersion));
+    LOG.info("read the model ver=" + jobModelVersion + " from " + keyBuilder.getJobModelPath(jobModelVersion));
```
Is it usually helpful to log the `processorId` in state transitions / reads like these? Could potentially be useful if there are multiple `StreamProcessor`s running?

nickpan47 replied:
Prefer what Jagadish said as well.

navina replied:
Yeah. I am going to change this. I think I already responded to @nickpan47 's comments before.

nickpan47 commented on samza-api/src/main/java/org/apache/samza/runtime/ProcessorIdGenerator.java:
```diff
@@ -0,0 +1,37 @@
+   * This value can be representative of its current executing environment. It can also be custom-managed by the user.
+   *
+   * <b>Note</b>: processorId has to be unique among the processors within a job. Processors can co-exists within the
+   * same JVM, as long as the identifiers are guaranteed to be unique.
```
Question: is the part to differentiate multiple processors in the same physical JVM process defined? i.e. did we define that a processorId include:
1) a cluster-specific way to identify the physical JVM process in a job
2) a defined way to differentiate multiple processors in the same physical JVM process

I think that we should make it explicit in the javadoc for this class.

navina replied:
@nickpan47 As I showed you in-person before, I added this to the design doc. I will include the specific details in the javadoc as well. Please wait for me to push the changes before reviewing it again. Thanks!

navina commented on samza-api/src/main/java/org/apache/samza/runtime/ProcessorIdGenerator.java:
```diff
@@ -0,0 +1,37 @@
+public interface ProcessorIdGenerator {
+  /**
+   * Generates a String representation to identify a single instance of StreamProcessor
+   * This value can be representative of its current executing environment. It can also be custom-managed by the user.
```
Will fix the typos

> Not sure if this is helpful: This value can be representative of its current executing environment. It can also be custom-managed by the user. I suspect it can be removed. You are already stating out the general contract for the ID. (in the following lines).

Why is not helpful though? If you think it is misleading, I can remove it. I will be updating this doc again based on some of @nickpan47 's feedback.

navina replied:
> s/StreamProcessor/{@link StreamProcessor}

ProcessorIdGenerator is in samza-api and StreamProcessor is in samza-core. I won't be able to make this change unless we move StreamProcessor to samza-api.

navina commented on samza-core/src/main/java/org/apache/samza/container/grouper/task/GroupByContainerCount.java:
```diff
@@ -225,10 +225,11 @@ private void validateTasks(Set<TaskModel> tasks) {
    * @param taskNamesToAssign     the list of tasks to assign to the containers.
    * @param containers            the containers (as {@link TaskGroup}) to which the tasks will be assigned.
    */
+  // TODO: Change logic from using int arrays to ordered array list
```
Oh.. That is a really old comment when I had not fully read through the implementation of the class. Let me fix the comment. Thanks!

jmakes replied:
The comment hasn't been removed yet.

navina replied:
Missed it before. Comment fixed!

navina commented:
> It "generates" some containerIds according to the count. It receives others from external sources like the task assignment manager. I was only suggesting that the latter should be validated s.t. we throw an exception rather than letting unexpected/undefined behavior occur.

Yeah. I agree we can test the latter. The place where you had added the comment threw me off. It should probably be in getPreviousContainers, where we readTaskAssignment.

> Also, I believe this is not equivalent to integer comparison.
In a sort scenario, for the same id's:
Integer comparison would yield [3, 8, 10]
String comparison would yield [10, 3, 8]
We should be certain that this will not affect any other code that uses ContainerModel as a comparator. e.g. for sort. If that code expects them to be in integer order for example, it could break. Or it could produce inconsistent behavior, different mappings of anything that is assigned to containers in a round-robin fashion.

I totally agree that comparison between Integer and String will result in different order. I am actually not even sure if "comparable" even makes sense anymore. I am trying to figure out where in our code we rely on "ordering" or "sorting" of ContainerModels. I couldn't find any use-case for this. 
In fact, I removed it and ran a build and it seems to work fine. However, that may not be sufficient as our test coverage is pretty poor :(

navina commented:
@jmakes @nickpan47 I have updated based on the feedback. Hopefully,I didn't miss any related to this SEP. Most recent changes:
* updated the documentation in ProcessorIdGenerator and ApplicationConfig
* Added validation in GroupByContainerCount
* Removed compareTo in ContainerModel

nickpan47 commented on samza-core/src/main/java/org/apache/samza/container/grouper/task/GroupByContainerCount.java:
```diff
@@ -211,7 +220,7 @@ private void validateTasks(Set<TaskModel> tasks) {
   private List<TaskGroup> createContainers(int startContainerId, int endContainerId) {
     List<TaskGroup> containers = new ArrayList<>(endContainerId - startContainerId);
     for (int i = startContainerId; i < endContainerId; i++) {
```
nit: seems like this is the reason why this Grouper does not work w/ non-integer processorIds? We probably should open a JIRA to fix it.

navina replied:
I think there is more than one place in this Grouper where the integer assumption is used. The intention was to re-write this grouper with the new standalone. I will make sure there is a JIRA for this. Thanks :)

nickpan47 commented on samza-core/src/main/java/org/apache/samza/container/grouper/task/GroupByContainerIds.java:
```diff
@@ -89,7 +89,7 @@ public GroupByContainerIds(int count) {
     Set<ContainerModel> containerModels = new HashSet<>();
     for (int i = 0; i < containerCount; i++) {
-      containerModels.add(new ContainerModel(containersIds.get(i), taskGroups[i]));
+      containerModels.add(new ContainerModel(containersIds.get(i), -1, taskGroups[i]));
```
Not sure why here we are using -1 and in line 79 we are using 'i'? Better to be consistent. If there is a reason to use either '-1' or 'i', please add a comment to explain.

navina replied:
We are setting it to -1 because there is no integer equivalent of the processorId anymore. 
I believe we are using "i" before just as index to the array of processorIds passed in the argument. 
I added a comment next to the -1 to clarify. Thanks!

nickpan47 commented on samza-core/src/main/java/org/apache/samza/coordinator/JobCoordinator.java:
```diff
@@ -55,12 +59,16 @@
-   * Returns the logical ID assigned to the processor
-   * It is up to the user to ensure that different instances of StreamProcessor within a job have unique processor ID.
-   * @return integer representing the logical processor ID
+   * Returns the logical ID assigned to the processor that is local to the instance of StreamProcessor.
```
nit: why is this called "logic"? I would think this is a "physical" ID since it is specific to the underlying environment already.

navina replied:
Yeah. This is a legacy comment. Let me fix it. Thanks for pointing it out.

nickpan47 commented on samza-core/src/main/java/org/apache/samza/runtime/UUIDGenerator.java:
```diff
@@ -0,0 +1,39 @@
+   * Generates a String representation to identify the processor instance
+   * This value can be representative of its current executing environment. It can also be custom-managed by the user.
+   * <p>
+   * <b>Note</b>: processorId has to be unique among the processors within a job
```
This does not follow the template we stated in the Javadoc (i.e. $x_$y format). It would better to be consistent with the documentation.

navina replied:
I have stated why the format exception is made in this use-case.

navina commented on samza-api/src/main/java/org/apache/samza/runtime/ProcessorIdGenerator.java:
```diff
@@ -0,0 +1,50 @@
+   *  processor instance within the JVM.</li>
+   * </ul>
+   *
+   * <b>Note</b>:
```
I clarified in the second point that when one processor exists per JVM, it should be fine to simply use "x" part to identify the processor. "Note" itself gives an example of how it can be achieved.

navina commented:
> No matter what implementation a JobCoordinator class takes, as long as it is loading ProcessorIdGenerator implementation, it should use the loader method and will perform this validation.

Our config validations are usually encapsulated in the corresponding "Config" class. Here, we are simply testing this for compatibility with existing streamprocessor users. Since it is temporary, I think it should be fine.
